From 4831ca3884c774a0817b2d96b479a13d12d5a63d Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 10:12:40 -0400
Subject: [fs] ext4: fix removing status extents in ext4_collapse_range()

Message-id: <1414059176-6044-17-git-send-email-lczerner@redhat.com>
Patchwork-id: 98287
O-Subject: [RHEL 7.1 PATCH 16/32] ext4: fix removing status extents in ext4_collapse_range()
Bugzilla: 1150171

BZ 1150171
https://bugzilla.redhat.com/show_bug.cgi?id=1150171

BREW 8149772
https://brewweb.devel.redhat.com/taskinfo?taskID=8149772

Upstream commit 2c1d23289bc2f7cfa358bc856b87a992dcb11ad5

 Currently in ext4_collapse_range() when calling ext4_es_remove_extent() to
 remove status extents we're passing (EXT_MAX_BLOCKS - punch_start - 1)
 in order to remove all extents from start of the collapse range to the
 end of the file. However this is wrong because we might miss the
 possible extent covering the last block of the file.

 Fix it by removing the -1.

 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Reviewed-by: Namjae Jeon <namjae.jeon@samsung.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 4a66aeb..47bb147 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -5449,7 +5449,7 @@ int ext4_collapse_range(struct inode *inode, loff_t offset, loff_t len)
  ext4_discard_preallocations(inode);
 
  ret = ext4_es_remove_extent(inode, punch_start,
-        EXT_MAX_BLOCKS - punch_start - 1);
+        EXT_MAX_BLOCKS - punch_start);
  if (ret) {
   up_write(&EXT4_I(inode)->i_data_sem);
   goto out_stop;
-- 
1.7.1